### PR TITLE
Add declaration to session after modifying a default PROTO parameter

### LINKED ISSUE #5026: Assertion failure with default PROTO parameter declaration
Webots crashes due to an assertion failure when opening the `derived_proto_with_template_default_parameter.wbt` test.
After the load the scene tree selection is updated to the `TemplateSolidCylinder` parameter node of the `ProtoWithTemplateDefaultParameter` PROTO.

`TemplateSolidCylinder` is correctly declared in the `ProtoWithTemplateDefaultParameter` PROTO file but not in the world file/scene tree level and causes the following assertion failure:

```
webots-bin: vrml/WbProtoManager.cpp:826: QString WbProtoManager::externProtoDeclaration(const QString&, bool) const: Assertion `false' failed.

Thread 1 "webots-bin" received signal SIGABRT, Aborted.
__pthread_kill_implementation (no_tid=0, signo=6, threadid=140737251429696) at ./nptl/pthread_kill.c:44
44	./nptl/pthread_kill.c: No such file or directory.
(gdb) bt
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140737251429696) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=140737251429696) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=140737251429696, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x00007ffff4bb1476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007ffff4b977f3 in __GI_abort () at ./stdlib/abort.c:79
#5  0x00007ffff4b9771b in __assert_fail_base
    (fmt=0x7ffff4d4c150 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=0x5555561b8e7b "false", file=0x5555561b8664 "vrml/WbProtoManager.cpp", line=826, function=<optimized out>)
    at ./assert/assert.c:92
#6  0x00007ffff4ba8e96 in __GI___assert_fail
    (assertion=0x5555561b8e7b "false", file=0x5555561b8664 "vrml/WbProtoManager.cpp", line=826, function=0x5555561b8e30 "QString WbProtoManager::externProtoDeclaration(const QString&, bool) const")
    at ./assert/assert.c:101
#7  0x0000555555dd1a4c in WbProtoManager::externProtoDeclaration(QString const&, bool) const (this=0x555556971fc0, protoName=..., formatted=true) at vrml/WbProtoManager.cpp:826
#8  0x0000555555d748d4 in WbNodeEditor::update() (this=0x5555565800e0) at scene_tree/WbNodeEditor.cpp:160
#9  0x0000555555d745fa in WbNodeEditor::edit(bool) (this=0x5555565800e0, copyOriginalValue=true) at scene_tree/WbNodeEditor.cpp:128
#10 0x0000555555f3ad33 in WbValueEditor::edit(WbNode*, WbField*, int) (this=0x5555565800e0, node=0x0, field=0x55555c16bb20, index=0) at scene_tree/WbValueEditor.cpp:79
#11 0x0000555555d7a4f2 in WbNodePane::edit(WbNode*, WbField*, int) (this=0x555556965f30, node=0x0, field=0x55555c16bb20, index=0) at scene_tree/WbNodePane.cpp:70
#12 0x0000555555c6c993 in WbFieldEditor::editField(WbNode*, WbField*, int) (this=0x5555566a6240, node=0x0, field=0x55555c16bb20, item=0) at scene_tree/WbFieldEditor.cpp:251
#13 0x0000555555e3776e in WbSceneTree::updateSelection() (this=0x55555693a550) at scene_tree/WbSceneTree.cpp:1143
```

### PR DESCRIPTION
**Pull request by stefaniapedrazzi**
Fixes #5026:
* [x] when modifying a PROTO parameter the PROTO declaration is added to the world EXTERNPROTO declarations
* [x] when displaying the PROTO URL (Add Node dialog and Node editor), use the node URL directly if the node is not found in the world declarations
* [x] remove declaration when resetting PROTO to default value

Tests:
* [x] open world with local PROTO not declared and select it
* [x] open world with a not declared PROTO node as default PROTO node parameter (not written in WBT file), and select it
* [x] modify default parameter, save world -> PROTO is added to the WBT file & no warnings are printed
* [x] reset default parameter, save world -> PROTO is removed from the WBT file & no warnings are printed
* [x] use different URL in world and PROTO file for default parameter:
  * [x] in case of default parameter  -> PROTO declaration should be used (after save and revert)
  * [x] in case of modified parameter  -> world declaration should be used (after save and revert)
* [ ] ...

### CODE REVIEW

**Comment by stefaniapedrazzi** [comment]
Currently, if the world has a different EXTERNPROTO URL than the PROTO the behavior is correct but only after saving and reverting the world.
However the conflicting message is printed in the console:
```
Conflicting declarations for '%1' are provided: %2 and %3, the first one will be used.
To use the other instead you will need to change it manually in the world file."
```

**Comment by stefaniapedrazzi** [comment]
Test suite is broken on the master branch. I tested locally and all the tests passed successfully.

**Comment by stefaniapedrazzi** [comment]
All the tests are now passing and this PR is ready for review.

File: src/webots/vrml/WbProtoManager.hpp [hunk 1]
```diff
@@ -19,6 +19,8 @@
 // Description: a class for managing a list of proto models
 //
 
+class WbField;
```
Comment (omichel) [inline src/webots/vrml/WbProtoManager.hpp:22]: Doesn't seem to be needed:
```suggestion
```

File: src/webots/nodes/WbBaseNode.cpp [hunk 2]
```diff
@@ -16,6 +16,7 @@
 #include "WbBasicJoint.hpp"
 #include "WbBoundingSphere.hpp"
 #include "WbDictionary.hpp"
+#include "WbField.hpp"
```
Comment (omichel) [inline src/webots/nodes/WbBaseNode.cpp:19]: Not needed.
```suggestion
```

File: src/webots/nodes/utils/WbNodeOperations.cpp [hunk 3]
```diff
@@ -36,6 +37,7 @@
 
 #include <QtCore/QCoreApplication>
 #include <QtCore/QFileInfo>
+#include <QtCore/QQueue>
```
Comment (omichel) [inline src/webots/nodes/utils/WbNodeOperations.cpp:40]: Not needed.
```suggestion
```

File: src/webots/nodes/utils/WbNodeOperations.cpp [hunk 4]
```diff
@@ -26,6 +26,7 @@
 #include "WbParser.hpp"
 #include "WbProject.hpp"
 #include "WbProtoManager.hpp"
+#include "WbProtoModel.hpp"
```
Comment (omichel) [inline src/webots/nodes/utils/WbNodeOperations.cpp:29]: Not needed.
```suggestion
```

**Review by omichel — CHANGES_REQUESTED** [review]

File: src/webots/nodes/utils/WbNodeOperations.cpp (hunk 4 above)
Comment (stefaniapedrazzi) [reply src/webots/nodes/utils/WbNodeOperations.cpp:29]: Yes, this is needed.
`WbProtoModel` class is used at line 468 and 475.

File: src/webots/nodes/WbBaseNode.cpp (hunk 2 above)
Comment (stefaniapedrazzi) [reply src/webots/nodes/WbBaseNode.cpp:19]: Yes, it is needed by the connect instruction.

**Review by omichel — CHANGES_REQUESTED** [review]

**Review by omichel — APPROVED** [review]
Thank you.
It looks good and seems to work nicely.

File: src/webots/vrml/WbProtoManager.cpp [hunk 5]
```diff
@@ -52,8 +52,6 @@ WbProtoManager::WbProtoManager() {
   mTreeRoot = NULL;
   mExternProtoCutBuffer = NULL;
 
-  mImportedFromSupervisor = false;
-
```
Comment (omichel) [inline src/webots/vrml/WbProtoManager.cpp:56]: This cause #5078.
It should be re-added otherwise the `mImportedFromSupervisor` is not initialized.

Comment (stefaniapedrazzi) [reply src/webots/vrml/WbProtoManager.cpp:56]: It seems it was introduced by mistake when merging with origin/master.
in any case it is not related to the fix.

**Review by omichel — APPROVED** [review]
Thank you.

**Pull request merged**